### PR TITLE
Migrate to new config

### PR DESCRIPTION
**Pull request by natemoo-re**
## Changes

- Implements [RFC0019](https://github.com/withastro/rfcs/blob/main/proposals/0019-config-finalization.md)
- This is a massive PR that touches a ton of files. Most changes are just renaming config, but some additional logic had to be moved around for:
  - `site` and `base`
  - `pages` being removed in favor of `srcDir`
- At one point, this PR attempted to auto-generate the correct config for users. Turns out this is pretty difficult, so I skipped it. Instead, I focused on providing a well typed `AstroUserConfig` object that should point users in the right direction.
  - If we detect a legacy config, we'll exit early and point users to our migration guide.

## Testing

Tests have all been updated

## Docs

Not yet

### CODE REVIEW

File: packages/astro/src/core/config.ts [hunk 1]
```diff
@@ -88,46 +130,33 @@ export const AstroConfigSchema = z.object({
...
-				.default('directory'),
-			legacyBuild: z.boolean().optional().default(false),
-			experimentalStaticBuild: z.boolean().optional().default(true),
-			experimentalSsr: z.boolean().optional().default(false),
-			drafts: z.boolean().optional().default(false),
+				.default('shiki'),
+			// TODO: add better type checking
+			shikiConfig: z.any().optional().default({}),
+			remarkPlugins: z.array(z.any()).optional().default([]),
+			rehypePlugins: z.array(z.any()).optional().default([]),
```
Comment (JuanM04) [inline packages/astro/src/core/config.ts:147]: Our markdown integration [doesn't distinguish](https://github.com/withastro/astro/blob/23b8ee2738076225da57ce5e3826373c8964717c/packages/markdown/remark/src/index.ts#L59-L60) Remark from Rehype plugins, so maybe it's better to have them in a single `plugins` array. What do you think?

File: packages/astro/src/core/config.ts [hunk 2]
```diff
@@ -43,40 +44,81 @@ async function resolvePostcssConfig(inlineOptions: any, root: URL): Promise<Post
...
+	'markdownOptions',
+	'buildOptions',
+	'devOptions',
+	'experimentalIntegrations'
+]);
+
 export const AstroConfigSchema = z.object({
-	projectRoot: z
+	adapter: z.object({ name: z.string(), hooks: z.object({}).passthrough().default({}) }).optional(),
+	logLevel: z
```
Comment (bholmesdev) [inline packages/astro/src/core/config.ts:62]: Nit: remove since `logLevel` isn't supported yet

File: packages/astro/test/test-utils.js [hunk 3]
```diff
@@ -69,7 +69,15 @@ export async function loadFixture(inlineConfig) {
 	}
 	// Load the config.
 	let config = await loadConfig({ cwd: fileURLToPath(cwd) });
-	config = merge(config, { ...inlineConfig, projectRoot: cwd });
+	config = merge(config, { ...inlineConfig, root: cwd });
+
+	// Note: the inline config doesn't run through config validation where these normalizations usually occur
```
Comment (bholmesdev) [inline packages/astro/test/test-utils.js:74]: Ah ha! Glad it was easy to re-create validation here. Still confused why we couldn't import core's `validateConfig` for this... but this seems like a fine stand-in.

**Review by bholmesdev — APPROVED** [review]
Small comment but looks migreat!

File: packages/astro/src/core/config.ts (hunk 1 above)
Comment (bholmesdev) [reply packages/astro/src/core/config.ts:147]: I second this if possible!

Comment (natemoo-re) [reply packages/astro/src/core/config.ts:147]: It actually does AFAICT. They're both loaded at the same time, but _when_ the plugins are applied is important. Remark plugins come first, then we convert Remark => Rehype, then we apply the Rehype plugins.

**Pull request merged**